### ISSUE TITLE
MTU Size for Protocol Version 3

### ISSUE THREAD

**Issue by justinjdickow** [smartdevicelink/sdl_ios#256]
Please see: https://github.com/smartdevicelink/protocol_spec/issues/4

**Comment by justinjdickow** [comment]
@Heath-FoMoCo @stefanek22 does this need to be completed for the 4.0.0 library release?

**Comment by stefanek22** [comment]
Yes, otherwise it will not be backwards compatible with the existing production work stream.

**Comment by joeljfischer** [comment]
For future reference on this issue: Currently there are constants in multiple places for 1024 or 512, depending on where you look. All of those will have to be changed to the version dependent system. 

We are doing the variable MTU like so: https://github.com/smartdevicelink/protocol_spec/pull/5

Does the v4 variable MTU size need to be approved on the protocol level, and developed on the lib level before v4 can roll out? I would assume so since lib v4.0.0 on iOS is claiming to be protocol v4 compatible. What are the next steps for getting it approved?

**Comment by justinjdickow** [comment]
@stefanek22 Okay then we need to make sure that this fix also makes it in as a hotfix on 4.0.0 in core.

**Comment by stefanek22** [comment]
Yes, definitely.

**Comment by justinjdickow** [comment]
@stefanek22 @Heath-FoMoCo 

Please confirm for SDL Core, Android, and iOS:

Protocol version 2: current MTU size
Protocol version 3: 128k
Protocol version 4: 128k
Protocol version ?: Dynamic MTU

If this is correct we can go ahead and confirm that it's going to be feasible for partners using 3 and 4.

**Comment by stefanek22** [comment]
If Protocol Version 4 cannot contain Dynamic MTU, then yes, it should remain the same as Protocol Version 3 (128K).

**Comment by justinjdickow** [comment]
Confirmed from partners that we can make the 1500 to 128k change in protocol v 3 and 4

**Comment by joeljfischer** [comment]
So the problem that I'm having with implementing this is:
* Where to put the method that calculates the MTU size? I chose to put it in the SDLProtocol because that's where the version is stored (It's also stored in the proxy it seems?).
* Putting it in the SDLProtocol class has issues:
  * Do I make it a class or instance method? I chose to make it an instance method because the version is held by the protocol, so it has to be used to calculate the MTU size.
    * If it's an instance method, how to classes that don't hold a reference to the SDLProtocol object use it? It seems that they cannot. This is currently only an issue for some tests on methods.
    * The two classes that currently need to reference it are SDLProxy (for putfile) and the protocol itself. The protocol has no problems, but the proxy, because of the abstraction to the SDLAbstractProtocol which it holds instead of the SDLProtocol itself, cannot reach this method.

The best solution to the SDLProxy problem would be to have it hold a reference directly to the SDLProtocol and remove the abstraction. This would also be manifested by removing the requirement to pass in a protocol when instantiating the proxy.

The best solution to the problem of referencing it from unrelated classes would be to have the proxy be a singleton, but that solution seems infeasible for the time being until a larger re-architecture of the high level API occurs. So I think the test will have to be altered or removed in some way. 

The test is for the SDLProtocolMessageDisassembler method (see issue #269)

**Comment by joeljfischer** [comment]
@adein @justinjdickow I wrote a novella above, but I could use some feedback.

**Comment by mrapitis** [comment]
@stefanek22 @justinjdickow

Please see my comment / suggestion below in regards to the 128k MTU comment previously made.

Protocol version 2: current MTU size
Protocol version 3: 128k
Protocol version 4: 128k
Protocol version >4: Dynamic MTU  -- Lowest possible support for 128k to protect existing proxy implementations for future hardware.  (This would mean all >V4 hardware would have to support at least 128K)  See the set version method below for details.  We could also choose to be more prudent and use a different (small) MTU value for anything above version 4.

```
private static final int V1_V2_MTU_SIZE = 1500;
private static final int V3_V4_MTU_SIZE = 131072;

public void setVersion(byte version) {
		this._version = version;

		if (version >= 4) {
			this._version = 4; //protect for future, proxy only supports v4 or lower
			HEADER_SIZE = 12;
			MAX_DATA_SIZE = V3_V4_MTU_SIZE; //versions 4 supports 128k MTU
			_headerBuf = new byte[HEADER_SIZE];
		} else if (version == 3) {
			this._version = 3;
			HEADER_SIZE = 12;
			MAX_DATA_SIZE = V3_V4_MTU_SIZE; //versions 3 supports 128k MTU
			_headerBuf = new byte[HEADER_SIZE];
		} else if (version == 2) {
			this._version = version;
			HEADER_SIZE = 12;
			MAX_DATA_SIZE = V1_V2_MTU_SIZE - HEADER_SIZE;
			_headerBuf = new byte[HEADER_SIZE];
		}
		else if (version == 1){
			HEADER_SIZE = 8;
			MAX_DATA_SIZE = V1_V2_MTU_SIZE - HEADER_SIZE;
			_headerBuf = new byte[HEADER_SIZE];			
		}
	}
```

**Comment by joeljfischer** [comment]
@mrapitis 
You're right, I didn't think about that, that a v3 proxy would interact with v5 hardware and require a 128k MTU. I don't think that's necessarily acceptable b/c of low end hardware we might want to support. We can:

1. Only every support 128k+
2. Drop our 128k in v3/4
3. Require developers that want to be on dynamic MTU hardware to have a newer library

Thoughts on those 3 options? It seems to me like (2) is our only answer right now if we're still committed to dual-backward-compatibility, which is a pretty crazy conclusion.

**Comment by justinjdickow** [comment]
@idamaster @stefanek22 @joeljfischer @joeygrover @mrapitis @adein We need to make a decision on this ASAP.

tldr; 128k hardcoded for v3 means all future hardware must support a minimum of 128k to maintain backwards compatibility with higher version apps

Recommendation from Livio team:

If head units that have higher versions than apps have to remain compatible, we should implement dynamic MTU. If we're doing dynamic we can make it a top priority to provide pull requests in iOS, Android, and Core.

If we can remove the dual backward compatibility, we can just go with hardcoded 128k right now.

**Comment by mrapitis** [comment]
+Heath
@Heath-FoMoCo 
I don't think we can remove the backwards compatibility because we need to protect for when an old proxy connects to a hu with a newer protocol version.

**Comment by joeljfischer** [comment]
Okay, so we need to get rid of 128k MTU until we have dynamic in order to account for this. Agreed?

**Comment by stefanek22** [comment]
Can we compromise by pushing forward with a 128k MTU for protocol v3 defined on SDL core, but the mobile library keeping the currently defined MTU size?  This way, certain libraries that require a higher MTU can make use of that, provided they commit to updating to newer mobile libraries once a dynamic MTU negotiation is implemented?  Having a higher MTU on the headunit side should not be an issue.

**Comment by Heath-FoMoCo** [comment]
Can we reexamine Markos's comment "We could also choose to be more prudent and use a different (small) MTU value for anything above version 4." (DEFAULT TO LOWER MTU SIZE).

After a quick discussion internally we thought this would be the best solution.  This would prevent an older proxy on a new low end head unit from failing.  If a partner fails to update to a newer proxy (with MTU negotiation) they would only be punishing themselves (slower performance) and not the SDL project (failed connection).

Thoughts?

**Comment by joeljfischer** [comment]
@Heath-FoMoCo @mrapitis 
Can you explain this further? I'm not quite sure what you mean by lower the default for version > 4. It seems to me that this wouldn't fix the issue. v3 (and maybe 4) proxies would still be trying to send 128k bytes per transmission, and if the default were lower on a head unit they were attempting to connect to there would be issues.The head unit thinks it's going to get 5k (random choice) bytes, then gets 128k bytes would have to handle that case. In other words, it seems like it wouldn't just be lower performance. That's what I'm thinking, but I'm probably missing something, can you explain a bit more?

**Comment by Heath-FoMoCo** [comment]
If we release a proxy that knows V3 and V4 of the proxy is 128K we can negotiate that.  If this same proxy connects and gets a V5 protcol it will not know if it supports 128K or is a lower end headunit which supports something lower, it connects using the 5K MTU.  This does not impact the module because anything the proxy isn't 100% sure can handle 128K it will send 5K.

Essentially we default the MTU to 5K until Protocol Version Negotiation determines without a shadow of a doubt something large can be handled by the head unit.

**Comment by mrapitis** [comment]
The updated code sample could look similar to:

```
private static final int V1_V2_MTU_SIZE = 1500;
private static final int V3_V4_MTU_SIZE = 131072;

public void setVersion(byte version) {
        this._version = version;

      if (version > 4) {
            this._version = 4; //protect for future, proxy only supports v4 or lower
            HEADER_SIZE = 12;
            MAX_DATA_SIZE = V1_V2_MTU_SIZE - HEADER_SIZE; //we don't know what size MTU is supported, default to lowest available
            _headerBuf = new byte[HEADER_SIZE];
        } else if (version == 4) {
            this._version = version; 
            HEADER_SIZE = 12;
            MAX_DATA_SIZE = V3_V4_MTU_SIZE; //versions 4 supports 128k MTU
            _headerBuf = new byte[HEADER_SIZE];
        } else if (version == 3) {
            this._version = version;
            HEADER_SIZE = 12;
            MAX_DATA_SIZE = V3_V4_MTU_SIZE; //versions 3 supports 128k MTU
            _headerBuf = new byte[HEADER_SIZE];
        } else if (version == 2) {
            this._version = version;
            HEADER_SIZE = 12;
            MAX_DATA_SIZE = V1_V2_MTU_SIZE - HEADER_SIZE;
            _headerBuf = new byte[HEADER_SIZE];
        } else if (version == 1){
            HEADER_SIZE = 8;
            MAX_DATA_SIZE = V1_V2_MTU_SIZE - HEADER_SIZE;
            _headerBuf = new byte[HEADER_SIZE];         
        }
    }
```

**Comment by joeljfischer** [comment]
Right, okay, I got ya. I see a few problems with that
1. We're excluding any and all lower end head units until we release dynamic MTU, which isn't ideal. 
2. It opens some complexity around testing. We're essentially creating a special case just for v3 -> v3 interaction with great performance, but the future would get worse. A v3 proxy will get great performance connected to a HU until that HU updates, then it will get much worse, until they update their proxy version (which most will hopefully do).
Proxy | HU
v2 -> anything = 1500
v3 -> v2 = 1500
v3 -> v3 = 128k
v3 -> >v3 = 5000 (or 1500 or whatever)
v4 -> v2 = 1500
v4 -> v3 = 128k
v4 -> v4 = dynamic (assuming dynamic is added in v4)

With that said, we're okay with this solution, we just wanted to understand the additional corner cases we're adding and make sure you guys are aware and good with it too.

**Comment by Heath-FoMoCo** [comment]
I don't understand #1: 
 >  We're excluding any and all lower end head units until we release dynamic MTU, which isn't ideal.

Are you saying that someone will release a lower end head unit that will have an MTU other than 1500 before SDL can support MTU negotiation?

>A v3 proxy will get great performance connected to a HU until that HU updates, then it will get much worse, until they update their proxy version (which most will hopefully do).

This is true: which I addressed in my previous comment.  The proxy *should* be released long before a head unit supports V5.  It will be up to the developer to be compliant or they will be hindering themselves.  Plus, the only apps this will affect are apps that require extremely large putfiles or video streaming.  A very small sample.

**Comment by joeljfischer** [comment]
Regarding 1, I just meant that if v3 head units are required to be 128k MTU, then a head unit that can't support 128k cannot use SDL until dynamic MTU support released.

**Comment by justinjdickow** [comment]
@joeljfischer Also need this update in iOS for 4.0.0

**Comment by justinjdickow** [comment]
@joeljfischer If you're doing this issue I'm going to go ahead and do the lock screen icon url implementation on iOS
